torch/csrc/autograd/python_cpp_function.cpp

```diff
@@ -227,6 +227,7 @@ PyTypeObject* _initFunctionPyTypeObject(
     const char* name,
     PyGetSetDef* function_properties,
     PyMethodDef* function_methods) {
+  type.ob_base = { PyObject_HEAD_INIT(nullptr) 0 };
```

Comment:
This is asked by Sam to actually initialize the header properly and make the objects immortal (since they are static) and this will init the refcount to 1 with cpython <3.12.
The incref here is removed but the ones for the objects created via addSubclass<>() are left since they call API that steal references.